Rust 1.19 regression, treebitmap 0.2.2

https://github.com/hroi/treebitmap

```
commit 67b8a1d591dd13a26b36519a92f5215a53c5c02e
Author: Hroi Sigurdsson <hroi@asdf.dk>
Date:   Mon Jan 23 16:15:28 2017 +0100

    travis
```

```
brian@ip-10-145-43-250:/mnt2/dev⟫ rustc +nightly -Vv
rustc 1.19.0-nightly (0418fa9d3 2017-06-04)
binary: rustc
commit-hash: 0418fa9d382a47d782cc1e195c14573be9c32095
commit-date: 2017-06-04
host: x86_64-unknown-linux-gnu
release: 1.19.0-nightly
LLVM version: 4.0

```

```

brian@ip-10-145-43-250:~/dev/slate/treebitmap⟫ cargo +nightly test
   Compiling treebitmap v0.2.2 (file:///mnt2/dev/slate/treebitmap)
error[E0311]: the parameter type `T` may not live long enough
   --> src/tree_bitmap/mod.rs:479:9
    |
479 | impl<T> IntoIterator for TreeBitmap<T> {
    |         ^^^^^^^^^^^^
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/tree_bitmap/mod.rs:479:9
    |
479 | impl<T> IntoIterator for TreeBitmap<T> {
    |         ^^^^^^^^^^^^

error[E0311]: the parameter type `T` may not live long enough
   --> src/tree_bitmap/mod.rs:479:9
    |
479 | impl<T> IntoIterator for TreeBitmap<T> {
    |         ^^^^^^^^^^^^
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/tree_bitmap/mod.rs:479:9
    |
479 | impl<T> IntoIterator for TreeBitmap<T> {
    |         ^^^^^^^^^^^^

error[E0311]: the parameter type `T` may not live long enough
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
273 | impl_ops!(Ipv4Addr);
    | -------------------- in this macro invocation
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
273 | impl_ops!(Ipv4Addr);
    | -------------------- in this macro invocation

error[E0311]: the parameter type `T` may not live long enough
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
273 | impl_ops!(Ipv4Addr);
    | -------------------- in this macro invocation
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
273 | impl_ops!(Ipv4Addr);
    | -------------------- in this macro invocation

error[E0311]: the parameter type `T` may not live long enough
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
274 | impl_ops!(Ipv6Addr);
    | -------------------- in this macro invocation
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
274 | impl_ops!(Ipv6Addr);
    | -------------------- in this macro invocation

error[E0311]: the parameter type `T` may not live long enough
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
274 | impl_ops!(Ipv6Addr);
    | -------------------- in this macro invocation
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
274 | impl_ops!(Ipv6Addr);
    | -------------------- in this macro invocation

error: aborting due to previous error(s)

error: Could not compile `treebitmap`.
Build failed, waiting for other jobs to finish...
error[E0311]: the parameter type `T` may not live long enough
   --> src/tree_bitmap/mod.rs:479:9
    |
479 | impl<T> IntoIterator for TreeBitmap<T> {
    |         ^^^^^^^^^^^^
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/tree_bitmap/mod.rs:479:9
    |
479 | impl<T> IntoIterator for TreeBitmap<T> {
    |         ^^^^^^^^^^^^

error[E0311]: the parameter type `T` may not live long enough
   --> src/tree_bitmap/mod.rs:479:9
    |
479 | impl<T> IntoIterator for TreeBitmap<T> {
    |         ^^^^^^^^^^^^
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/tree_bitmap/mod.rs:479:9
    |
479 | impl<T> IntoIterator for TreeBitmap<T> {
    |         ^^^^^^^^^^^^

error[E0311]: the parameter type `T` may not live long enough
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
273 | impl_ops!(Ipv4Addr);
    | -------------------- in this macro invocation
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
273 | impl_ops!(Ipv4Addr);
    | -------------------- in this macro invocation

error[E0311]: the parameter type `T` may not live long enough
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
273 | impl_ops!(Ipv4Addr);
    | -------------------- in this macro invocation
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
273 | impl_ops!(Ipv4Addr);
    | -------------------- in this macro invocation

error[E0311]: the parameter type `T` may not live long enough
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
274 | impl_ops!(Ipv6Addr);
    | -------------------- in this macro invocation
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
274 | impl_ops!(Ipv6Addr);
    | -------------------- in this macro invocation

error[E0311]: the parameter type `T` may not live long enough
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
274 | impl_ops!(Ipv6Addr);
    | -------------------- in this macro invocation
    |
    = help: consider adding an explicit lifetime bound for `T`
    = note: the parameter type `T` must be valid for the empty lifetime...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/lib.rs:259:17
    |
259 |         impl<T> IntoIterator for IpLookupTable<$addr_type, T> {
    |                 ^^^^^^^^^^^^
...
274 | impl_ops!(Ipv6Addr);
    | -------------------- in this macro invocation

error: aborting due to previous error(s)

error: build failed
```

cc @hroi

--- a/src/librustc/infer/region_inference/mod.rs
+++ b/src/librustc/infer/region_inference/mod.rs
@@ -1200,6 +1200,13 @@ impl<'a, 'gcx, 'tcx> RegionVarBindings<'a, 'gcx, 'tcx> {
         for verify in self.verifys.borrow().iter() {
             debug!("collect_errors: verify={:?}", verify);
             let sub = normalize(self.tcx, var_data, verify.region);
+
+            // This was an inference variable which didn't get
+            // constrained, therefore it can be assume to hold.
+            if let ty::ReEmpty = *sub {
+                continue;
+            }
+
             if verify.bound.is_met(region_rels, var_data, sub) {
                 continue;
             }

--- a/src/test/run-pass/issue-42467.rs
+++ b/src/test/run-pass/issue-42467.rs
@@ -1,0 +1,32 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+struct Foo<T>(T);
+
+struct IntoIter<T>(T);
+
+impl<'a, T: 'a> Iterator for IntoIter<T> {
+    type Item = ();
+
+    fn next(&mut self) -> Option<()> {
+        None
+    }
+}
+
+impl<T> IntoIterator for Foo<T> {
+    type Item = ();
+    type IntoIter = IntoIter<T>;
+
+    fn into_iter(self) -> IntoIter<T> {
+        IntoIter(self.0)
+    }
+}
+
+fn main() {}

